pkg/app/registry.go

```diff
@@ -0,0 +1,178 @@
+package app
+
+import (
+	"bufio"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	kotsv1beta1 "github.com/replicatedhq/kots/kotskinds/apis/kots/v1beta1"
+	"github.com/replicatedhq/kots/pkg/crypto"
+	"github.com/replicatedhq/kots/pkg/rewrite"
+	"github.com/replicatedhq/kotsadm/pkg/kotsutil"
+	"github.com/replicatedhq/kotsadm/pkg/logger"
+	"github.com/replicatedhq/kotsadm/pkg/persistence"
+	"go.uber.org/zap"
+)
+
+type RegistrySettings struct {
+	Hostname    string
+	Username    string
+	HasPassword bool
+	Namespace   string
+}
+
+func UpdateRegistry(appID string, hostname string, username string, password string, namespace string) error {
+	logger.Debug("updating app registry",
+		zap.String("appID", appID))
+
+	cipher, err := crypto.AESCipherFromString(os.Getenv("API_ENCRYPTION_KEY"))
+	if err != nil {
+		return errors.Wrap(err, "failed to create aes cipher")
+	}
+
+	passwordEnc := base64.StdEncoding.EncodeToString(cipher.Encrypt([]byte(password)))
+
+	db := persistence.MustGetPGSession()
+	query := `update app set registry_hostname = $1, registry_username = $2, registry_password_enc = $3, namespace = $4 where id = $5`
+	_, err = db.Exec(query, hostname, username, passwordEnc, namespace, appID)
+	if err != nil {
+		return errors.Wrap(err, "failed to update registry settings")
+	}
+
+	return nil
+}
+
+// RewriteImages will use the app (a) and send the images to the registry specified. It will create patches for these
+// and create a new version of the application
+func (a App) RewriteImages(hostname string, username string, password string, namespace string, configValues *kotsv1beta1.ConfigValues) error {
+	if err := SetTaskStatus("image-rewrite", "Updating registry settings", "running"); err != nil {
+		return errors.Wrap(err, "failed to set task status")
+	}
+
+	finishedCh := make(chan struct{})
+	defer close(finishedCh)
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Second):
+				if err := UpdateTaskStatusTimestamp("image-rewrite"); err != nil {
+					logger.Error(err)
+				}
+			case <-finishedCh:
+				return
+			}
+		}
+	}()
+
+	var finalError error
+	defer func() {
+		if finalError == nil {
+			if err := ClearTaskStatus("image-rewrite"); err != nil {
+				logger.Error(err)
+			}
+		} else {
+			if err := SetTaskStatus("image-rewrite", finalError.Error(), "failed"); err != nil {
+				logger.Error(err)
+			}
+		}
+	}()
+
+	// get the archive and store it in a temporary location
+	appDir, err := GetAppVersionArchive(a.ID, a.CurrentSequence)
+	if err != nil {
+		finalError = err
+		return errors.Wrap(err, "failed to get app version archive")
```

Comment:
Don't change this, but I think better pattern is to declare `finalError` in the function declaration and then here:
```
finalError = errors.Wrap(err, "failed to get app version archive")
return
```